server/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java

```diff
@@ -104,7 +104,9 @@ private void innerRun() throws Exception {
         final SearchPhaseController.ReducedQueryPhase reducedQueryPhase = resultConsumer.reduce();
         // Usually when there is a single shard, we force the search type QUERY_THEN_FETCH. But when there's kNN, we might
         // still use DFS_QUERY_THEN_FETCH, which does not perform the "query and fetch" optimization during the query phase.
-        final boolean queryAndFetchOptimization = queryResults.length() == 1 && context.getRequest().hasKnnSearch() == false;
+        final boolean queryAndFetchOptimization = queryResults.length() == 1
+            && context.getRequest().hasKnnSearch() == false
+            && reducedQueryPhase.rankCoordinatorContext() == null;
```

Comment:
This is specifically required for multiple queries. It wasn't missing before.

Reply:
I think what confuses me is that we don't look for the multiple queries but rather for the rank coordinator context. This could use a comment to explain why, or maybe the conditional could be shared in a single place, assuming it is always exactly the same.

Reply:
This is removed.

Reply:
I realized I didn't answer this correctly. This is required specifically for ranking. If we have a single query and multiple knn searches, those become a boolean query and it's fine to shortcut.